source/sdk/lang/Exception.ooc

```diff
@@ -1,397 +1,395 @@
-        }
-    }
-
-    /** Class which threw the exception. May be null */
-    origin: Class
-
-    /** Message associated with this exception. Printed when the exception is thrown. */
-    message : String
-
-    /**
-     * Create an exception
-     *
-     * @param origin The class throwing this exception
-     * @param message A short text explaning why the exception was thrown
-     */
-    init: func  (=origin, =message)
-
-    /**
-     * Create an exception
-     *
-     * @param message A short text explaning why the exception was thrown
-     */
-    init: func ~noOrigin (=message) {
-        init(null, message)
-    }
-
-    free: override func {
-      while (!this backtraces isEmpty) {
-        backtrace := this backtraces pop() as Backtrace
-        if (backtrace)
-          backtrace free()
-      }
-      this backtraces free()
-      super()
-    }
-
-    /**
-     * @return the exception's message, nicely formatted
-     */
-    formatMessage: func -> String {
-        if(origin)
-            "[%s in %s]: %s\n" format(class name toCString(), origin name toCString(), message ? message toCString() : "<no message>" toCString())
-        else
-            "[%s]: %s\n" format(class name toCString(), message ? message toCString() : "<no message>" toCString())
-    }
-
-    /**
-     * Print this exception, with its origin, if specified, and its message
-     */
-    print: func {
-        printMessage()
-        printBacktrace()
-    }
-
-    /**
-     * Print just the message
-     */
-    printMessage: func {
-        fprintf(stderr, "%s", formatMessage() toCString())
-    }
-
-    /**
-     * Throw this exception
-     */
-    throw: func {
-        _cleanupStackFrames()
-        _setException(this)
-        addBacktrace()
-        if(!_hasStackFrame()) {
-            version (windows) {
-                if (IsDebuggerPresent()) {
-                    // trigger a break point here, debugger will like that!
-                    printMessage()
-                    DebugBreak()
-                } else {
-                    // print the backtrace ourselves
-                    print()
-                }
-                exit(1)
-            }
-            version (!windows) {
-                printMessage()
-                abort()
-            }
-        } else {
-            frame := _takeStackFrame()
-            _markStackFrame(frame)
-            frame@ buf longJmp(_EXCEPTION)
-        }
-    }
-
-    /**
-     * Rethrow this exception.
-     */
-    rethrow: func {
-        throw()
-    }
-
-    getCurrentBacktrace: static func -> String {
-        result: String
-        h := BacktraceHandler get()
-        bt := h backtrace()
-        if (bt) {
-            result = h backtraceSymbols(bt)
-            bt free()
-        } else
-          result ="[no backtrace] use a debugger!\n"
-        result
-    }
+	backtraces := Stack<Backtrace> new()
+
+	addBacktrace: func {
+		bt := BacktraceHandler get() backtrace()
```

Comment:
Since I'm already blaming you for stuff you didn't do - `backtrace` and `handler` here and in the next function.